Паньшин Максим

makspanshin opened the pull request.

makspanshin committed:
Create project "TagsCloudVisualization"

makspanshin committed:
Create UnitTest project

makspanshin committed:
Update systems file

makspanshin committed:
Create Spiral class

Класс реализует спираль Архимеда.

makspanshin committed:
Implementation of the CircularCloudLayouter class

makspanshin committed:
Testing TagsCloudVisualization

makspanshin committed:
Delete UnitTest1.cs

makspanshin committed:
Update TagsCloudVisualization.csproj

makspanshin committed:
Update Spiral.cs

makspanshin committed:
Create PainterOfRectangles.cs

makspanshin committed:
Update CircularCloudLayouter.cs

makspanshin committed:
Create README.md

makspanshin committed:
Update README.md

makspanshin committed:
Implementation of the PainterOfRectangles

Класс для отрисовки CircularCloudLayouter

razor2651 commented on cs/TagsCloudVisualization/TagsCloudVisualization.csproj:
```diff
@@ -0,0 +1,57 @@
+    <OutputType>Exe</OutputType>
+    <RootNamespace>TagsCloudVisualization</RootNamespace>
+    <AssemblyName>TagsCloudVisualization</AssemblyName>
+    <TargetFrameworkVersion>v4.8</TargetFrameworkVersion>
```
а чо почему либа на framework 4,8, а тесты уже на net core?
ну тип ошибкой не назвать но поясни почему так?
да и может все таки единообразно сделаешь?

makspanshin replied:
Да как то не задумывался, всегда использовал framework. Исправлю

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,67 @@
+            if (rectangleSize.Width <= 0 || rectangleSize.Height <= 0)
+            {
+                throw new ArgumentException(
+                    $"Не корректные размера прямоугольника. Высота: {rectangleSize.Height} ," +
```
не корректные... кто?

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,67 @@
+                    $" Ширина: {rectangleSize.Width}");
+            }
+
+            if (rectangleSize.Width < 0 || rectangleSize.Width < 0)
```
зачем тебе два почти одинаковых ифа?
1) кажется тут бага ибо в одном только ширина проверяется а в другом и то и другое
2) тут очевидно дублирование

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,67 @@
+            }
+
+            Rectangle currentRectangle;
+            if (rectangles.Count == 0)
```
костыль: ты будешь проверять это условие каждый раз после добавления первого прямоугольника, нельзя ли сделать по проще?

razor2651 commented on cs/TagsCloudVisualization/PainterOfRectangles.cs:
```diff
@@ -0,0 +1,31 @@
+
+            bmp.Save(filename);
+
+            bmp.Dispose();
```
есть такой синтаксический сахар: `using`
посмотри что это такое, может найдется и тут применение

razor2651 commented on cs/TagsCloudVisualization/Program.cs:
```diff
@@ -0,0 +1,15 @@
+﻿using System;
```
лишнее убираем

razor2651 commented on cs/TagsCloudVisualization/TagsCloudVisualization.csproj:
```diff
@@ -0,0 +1,57 @@
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <ProjectGuid>{954A4E88-1549-4558-9046-E83D1F59EA33}</ProjectGuid>
+    <OutputType>Exe</OutputType>
```
ну это не правда, тебе тут не нужна exe

razor2651 commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,36 @@
+
+        public Point CalcPointSpiral()
+        {
+            var p = Step * countPoints / TwoPi;
```
что такое p?

makspanshin replied:
p - это полярный радиус. Исправлю на более понятное название

razor2651 commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,36 @@
+        public Point CalcPointSpiral()
+        {
+            var p = Step * countPoints / TwoPi;
+            var x = (int)(p * Math.Cos(countPoints + 1)) + center.X;
```
почему +1?

makspanshin replied:
Моя ошибка, там не должно быть `+1`

razor2651 commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,36 @@
+            var x = (int)(p * Math.Cos(countPoints + 1)) + center.X;
+            var y = (int)(p * Math.Sin(countPoints + 1)) + center.Y;
+
+            countPoints++;
```
countPoints что это?
насколько мне известно в уровнении спирали там как то играют роль такие вещи как радиус... что за непонятные переменные?

makspanshin replied:
Исправлю на `nextAngle`. Это угол следующей расчетной точки

razor2651 commented on cs/TagsCloudVisualization/PainterOfRectangles.cs:
```diff
@@ -0,0 +1,31 @@
+            pictSize = sizeOfPicture;
+        }
+
+        public void CreateImageOfRectangles(List<Rectangle> rectangles,string filename = "rectangles")
```
давай как то в одном стиле выдерживать код, тут пробел куда то пропал между аргументами

razor2651 replied:
какое-то дублирование везде в названиях
класс `PainterOfRectangles`
метод `CreateImageOfRectangles`
ничего не смущает?

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,67 @@
+        private Point nextPoint;
+        private List<Rectangle> rectangles;
+        private readonly Spiral spiral;
+        public List<Rectangle> Rectangles => rectangles;
```
а чо зачем нам тут публичное свойство и еще приватное поле?

makspanshin replied:
Подумал о будущем. Вдруг придется добавлять логику перед тем как передавать коллекцию прямоугольников

razor2651 replied:
понятно что нужно писать код так чтобы он был расширяем, но в разумных пределах, когда ты например понимаешь какое расширение возможно в перспективе исходя из предметной области, но и даже так без фанатизма

пример: есть асинхронный сервис отчетов, он умеет выполнять задачу, которая собирает данные и печает в txt, очевидно что в перспективе тут возможен какой-то чейнинг, например сначала собрать данныые, потом асинхронно напечатать в другом сервисе, который например делает pdf, т.е. появится последовательное выполнение подзадач, илил например собирать данные можно параллельно, тогда появляются параллельные асинхронные подзадачки которые сервис может выполнять

исходя из этого можно сразу заложить архитектуру такую чтобы условный отчет собирался из мелких кирпичиков, каждый был ответственен за свою мелкую задачу
ну т.е. есть какое-то понимание того как это может развиваться, и способствовало более простой инфраструктуре, там же где это пока не нужно, не понятно как может быть расширено, или тем более это не инфраструктурное изменение (ты если захочешь выделить свойство с backing field то ты не будешь менять интерфейс и будешь менять его конкретную реализацию) то такое лучше не делать

razor2651 replied:
проблема снова появилась, писал новый коммент, оставляю не resolved

razor2651 commented on cs/TagsCloudVisualizationUnitTest/CircularCloudLayouterTest.cs:
```diff
@@ -0,0 +1,44 @@
+        [TestCase(-10, -10, TestName = "Negative size")]
+        [TestCase(0, 0, TestName = "Size is Empty")]
+        [TestCase(10, 0, TestName = "Weight = zero")]
+        public void PutNextRectangleTest_ShouldBeThrowWhen(int h, int w)
```
полное название теста это не только имя метода и имя кейса
сюда так же входит имя класса
`TagsCloudVisualizationTest.PutNextRectangleTest_ShouldBeThrowWhen.Negative size`
тут 2 раза test повторяется не нужный, без него читается прям как текст

razor2651 replied:
h... w... что?

razor2651 commented on cs/TagsCloudVisualizationUnitTest/CircularCloudLayouterTest.cs:
```diff
@@ -0,0 +1,44 @@
+        }
+
+        [TestCase(10, 10, TestName = "Usual param")]
+        public void PutNextRectangleTest_ShouldNotThrowWhen(int h, int w)
```
утверждение: `бесполезный тест`
но т.к. он написан, то давай общаться
зачем этот тест?
что он проверяет?
не проверяется ли то же самое в других тестах?

makspanshin replied:
В предыдущем тесте  я там проверил ввод некорректных данных, а в этом тесте я проверяю не будет ли исключение с корректными значениями. 
Здесь мы явно проверяем данный метод. В других тестах, если этот метод будет задействован, то это уже не явная проверка. 
Вот такая у меня была логика. Я за явную проверку :)

razor2651 replied:
смотри тут есть несколько поинтов
1) когда ты смотришь на тесты то видишь это как какую-то спецификацию, и видя такой тест в котором проверяются на валидность конкретные параметры возникает вопрос а почему именно такие? это какое-то пограничное значение?
2) на самом деле внутри метода ты не проверяешь ни одну ветвь кода, если я в ифах уберу throw то тест останется зеленым
3) если ты в других тестах при использовании метода не пишешь try catch (что на самом деле плохо) то ты и так проверяешь свой not throw, потому что если кинется ошибка, то ты получишь красный тест, да, это не так явно как когда ты явно прописываешь это, но если бы была какая-то проверка результата то ок (например усновно основная  задача метода - кинуть ошибку или нет), а иначе можно на все подряд писать not throw

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,67 @@
+            return currentRectangle;
+        }
+
+        public bool IsIntersects(Rectangle rectangle)
```
вообще это внутренняя логика, почему это вынесено в интерфейс?

makspanshin replied:
Специально сделано для тестирования.

razor2651 replied:
т.е. тебя не смущает что для того чтобы протестить класс, ты открываешь наружу внутренности?

makspanshin replied:
Погуглил про тестирование приватных методов, идеального решение нет. Поэтому предлагаю: 
1) В тестовом классе реализовать метод `IsIntersects(Rectangle rectangle) `, так как у меня есть доступ к коллекции прямоугольников, ничего сложного тут не будет. 
2) Забить на приватные методы :) И считаю это будет не хорошо и не плохо так как идеального решение нет

makspanshin replied:
И все равно приватные методы будут проверяется не явно в методах интерфейса

razor2651 replied:
я не считаю что тут правильного решения нету
если ты хочешь проверить какую то приватную логику то кажется что эта логика слишком сложная и необходимо инкапсулировать это в отдельный класс и принимать эту логику как параметр в необходимом классе, тогда тот класс можно будет протестировать
что считать сложным - сложный вопрос, но кажется все должно исходить из здравого смысла

плюс ко всему ты на самом деле должен тестировать только публичный интерфейс конкретнйо реализации, потому что пользователь твой либы все равно ничего не будет знать про внутреннее устройство твоего метода, ему не нужны ни спеки по твоим внутренним методам ни чего либо еще

> И все равно приватные методы будут проверяется не явно в методах интерфейса

это не правда, тестируешь ты публичный интерфейс, то что в реализации используется какой-то метод - ок, но он может и не использоваться, он может быть а может и не быть, ты обязан гарантировать работу только внешних интерфейса а не каких то внутренних кишок
конкретный кейс может и остаться, но он должен тестировать интерфейс

razor2651 replied:
оставляю не resolved т.к. появились новые ошибки по этой же теме

razor2651 commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,36 @@
+            countPoints = 0;
+        }
+
+        public Point CalcPointSpiral()
```
Calc?
[Canadian Association of Lutheran Congregations?](https://en.wikipedia.org/wiki/Canadian_Association_of_Lutheran_Congregations)

razor2651 commented on cs/TagsCloudVisualizationUnitTest/CircularCloudLayouterTest.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace TagsCloudVisualizationUnitTest
+{
+    public class TagsCloudVisualizationTest
```
Требования к тестам:

- Все требования должны вычитываться из твоих модульных тестов.
- Ревьюеру должно быть легко по тесту понять проверяемое требование.

пока что тестов слишком мало чтобы говорить о каких-то требованиях вообще

makspanshin replied:
Можно добавить тесты:
1) На плотность
2) У всех прямоугольников должны быть уникальные координаты
3) Тест на скорость работы

razor2651 replied:
помимо укладчика в проекте дофига кода, он не протестирован, не вижу чтобы велась работа по tdd, а это была основная тема предыдущего занятия

makspanshin committed:
Update Spiral.cs

Изменил название переменных
Удалил баг "countPoints + 1"

makspanshin committed:
Update PainterOfRectangles.cs

Переписал с использованием using

makspanshin committed:
Update PainterOfRectangles.cs

Изменил название метода

makspanshin committed:
Update CircularCloudLayouter.cs

Удалил не нужный участок кода

makspanshin committed:
Update CircularCloudLayouter.cs

Рефакторинг

makspanshin committed:
Add new test

Тест проверяет на уникальность координат, то есть не должно быть прямоугольников с одинаковыми координатами

makspanshin committed:
Update CircularCloudLayouterTest.cs

Рефакторинг

makspanshin committed:
Update CircularCloudLayouter.cs

makspanshin committed:
Create GeneratorOfRectangles.cs

makspanshin committed:
Create SquareRectangle.cs

makspanshin committed:
Update TagsCloudVisualization.csproj

makspanshin committed:
Update CircularCloudLayouterTest.cs

makspanshin committed:
Update CircularCloudLayouterTest.cs

Добавил метод создание папки для снапшотов проваленных тестов

makspanshin committed:
Add Debug porject

makspanshin committed:
Net.Framework to Net.Core

makspanshin committed:
Обновил скриншоты

makspanshin committed:
Delete Program.cs

makspanshin committed:
Update GeneratorOfRectangles.cs

makspanshin committed:
Update PainterOfRectangles.cs

makspanshin committed:
Update TagsCloudVisualizationUnitTest.csproj

makspanshin committed:
Update CircularCloudLayouterTest.cs

Сделал 3 задание
Отрефакторил

makspanshin committed:
Update CircularCloudLayouter.cs

 Добавил метод CorrectLocationRectangle. Метод корректирует расположение прямоугольника, делает его более плотнее

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,136 @@
+            {
+                throw new ArgumentException(
+                    $"Не корректные размеры прямоугольника. Высота: {rectangleSize.Height} ," +
+                    $" Ширина: {rectangleSize.Width}");
```
строки лучше не конкатенировать, пусть они даже выбиваются за 120 символов

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,136 @@
+
+            Rectangles.Add(CorrectLocationRectangle(currentRectangle));
+
+            squareRectangles += currentRectangle.Square();
```
что это такое и зачем?

makspanshin replied:
При добавлении прямоугольника, я сразу вычисляю площадь облака, чтобы потом не  проходить по всей коллекции. 
Считаю полезный параметр, может понадобится  при отрисовки.

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,136 @@
+            centrPoint = center;
+            nextPoint = centrPoint;
+            Rectangles = new List<Rectangle>();
+            spiral = new Spiral(centrPoint);
```
вот тут у тебя зависимость строится внутри укладчика
спойлерить особо не буду, у вас еще будет занятие на инверсию зависимостей, контейнеры и вот это вот все
пока самый простой пример лишь приведу:
вот у тебя появится другая реализация спирали, тебе уже чтобы заиспользовать ее придется редактировать код и пересобирать приложение, это совсем не гибко, такие зависимости надо принимать параметрами

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,136 @@
+
+namespace TagsCloudVisualization
+{
+    public class CircularCloudLayouter
```
давай явно выделим интерфейс

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,136 @@
+        private readonly Spiral spiral;
+        private double squareRectangles;
+        public readonly List<Rectangle> Rectangles;
+        public double SquareRectangles => squareRectangles;
```
мы уже общались на эту тему, и вот опять

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,136 @@
+        private Point nextPoint;
+        private readonly Spiral spiral;
+        private double squareRectangles;
+        public readonly List<Rectangle> Rectangles;
```
опять же, зачем это public?

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,136 @@
+            return Rectangles.Any(rect => rect.IntersectsWith(rectangle));
+        }
+
+        public double Radius()
```
КАРАУЛ!
что это такое все?
зачем это все здесь?

мы уже обшались на тему того что ты открываешь интерфейс и добавляешь методы только ради тестирования, почему опять эти ошибки?
это нарушение инкапсуляции

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,136 @@
+            if (Rectangles == null)
+                throw new Exception();
+
+            var lastPoint = Rectangles.Last().Location; // последнняя точка самая дальняя от центра
```
1) из кода и так понятно что тут происходит - значит комментарий бесполезный
2) коммент не верный и вводит в заблуждение, это совершенно не факт что самая удаленная точка
3) комментарии в коде - зло, и свидетельсво грязного и непонятного кода, но в данном случае это просто мусор

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,136 @@
+        public double Radius()
+        {
+            if (Rectangles == null)
+                throw new Exception();
```
ммм... очень понятно что случилось

razor2651 commented on cs/TagsCloudVisualization/PainterOfRectangles.cs:
```diff
@@ -0,0 +1,29 @@
+
+            foreach (var rectangle in rectangles)
+            {
+                graphics.DrawRectangle(new Pen(Color.Blue, .5f), rectangle);
```
опять не освобождаешь ресурсы

razor2651 commented on cs/TagsCloudVisualization/PainterOfRectangles.cs:
```diff
@@ -0,0 +1,29 @@
+                graphics.DrawRectangle(new Pen(Color.Blue, .5f), rectangle);
+            }
+
+            bmp.Save(filename);
```
по хорошему эту штуку нужно бы выделить
у тебя класс - рисователь, а занимается он еще и сохранением
а если надо сохранять не в файл, а отдать по апи? а если в базу сохранить?

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,136 @@
+                    rectangle.Location = newPoint;
+                }
+
+                newPoint.Y++;
```
поясни вот тут логику, ты в цикле двигаешь в одну сторону, а потом когда цикл завершился один раз двигаешь обратно, не понимаю

makspanshin replied:
Костыль, выходит из цикла, когда пересечется с другим треугольником, поэтому я возвращаю его. Метод надо переписать

razor2651 replied:
ты делаешь продукт который будешь развивать в долгосрочной перспективе (спойлер: на этом работа с кодом не заканчивается, еще будет несколько задач где надо будет развивать идеи твоего подхода)
и на первом же этапе ты относишься к костылям абсолютно положительно, тебя совершенно не смущает их наличие, написание сложного кода который не понятен

как ты будешь развивать свое решение через 2 недели, через месяц, ты вспомнишь что ты тут делал?
по хорошему от костыля надо избавляться

razor2651 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,136 @@
+                newPoint.Y++;
+                rectangle.Location = newPoint;
+            }
+            else if (oldPoint.Y < centrPoint.Y)
```
а здесь не надо сравнить newPoint.Y != centrPoint.Y ?
и вообще тут дублирования слишко много по сути, нельзя ли как то избавиться от этого?

кажется что оба метода идентичны, тебе достаточно передать туда координату а не весь прямоугольник и с чем сравнивать (можно даже через func передать просто экспрешн) а потом просто двигать потихоньку
ну это просто идея
там кажется только мешать будет isintersect, но возможно эту проверку можно на уровень выше поднять

в общем давай попробуй, если не будет получаться, приходи, будем вместе думать

razor2651 replied:
не исправлено

razor2651 commented on cs/DebugProject/Program.cs:
```diff
@@ -0,0 +1,30 @@
+using System.Drawing;
+using TagsCloudVisualization;
+
+namespace DebugProject
```
круто что выделил отдельный проект под такие вещи, хочу понять мотивацию, есть ли понимание причин зачем так делать или это стечение обстоятельств и необходимо это рассказать

makspanshin replied:
Чтобы посмотреть, что получилось :) да и тебе потестить либу.

razor2651 replied:
эх... нет понимания((

безусловно желание хорошее но все таки тут я ожидал другого ответа
что если я зхаочу подключить твою либу к себе в вебапи? я тогда подключу вместе с генерилкой облака весь твой зоопарк с UI, по хорошему у тебя должны быть отделены сборки связанные с UI и с бизнес логикой

razor2651 commented on cs/TagsCloudVisualization/PainterOfRectangles.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace TagsCloudVisualization
+{
+    public class PainterOfRectangles
```
интерфейсы...

razor2651 commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,36 @@
+namespace TagsCloudVisualization
+{
+    /// <summary>
+    /// Класс реализует  спираль Архимеда
```
спасибо, я и так понял
мусор

razor2651 commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,36 @@
+    /// <summary>
+    /// Класс реализует  спираль Архимеда
+    /// </summary>
+    public class Spiral
```
интерфейсы выделяем

razor2651 commented on cs/TagsCloudVisualizationUnitTest/CircularCloudLayouterTest.cs:
```diff
@@ -0,0 +1,131 @@
+        [OneTimeSetUp]
+        public void CreateFolderForFailedTests()
+        {
+            if (!Directory.Exists(PathFolderFailedTest))
```
проверка бесполезная, CreateDirectory не кинет ошибку если папка создана, от того и OneTimeSetUp лишний

razor2651 commented on cs/TagsCloudVisualization/PainterOfRectangles.cs:
```diff
@@ -0,0 +1,29 @@
+            pictSize = sizeOfPicture;
+        }
+
+        public void CreateImage(List<Rectangle> rectangles, string filename = "rectangles")
```
а что если прямоугольники не влезут в изображение?
(поступить тут можно по разному, например кидать понятную ошибку, либо ресайзить изображение, а может еще что-то, какое-то конкретное решение не требую, но оно должно быть)

razor2651 commented on cs/TagsCloudVisualizationUnitTest/CircularCloudLayouterTest.cs:
```diff
@@ -0,0 +1,131 @@
+        public void InitCircularCloudLayouter()
+        {
+            circularCloudLayouter = new CircularCloudLayouter(new Point(500, 500));
+            painterOfRectangles = new PainterOfRectangles(new Size(1000, 1000));
```
тебе это нужно только на teardown

razor2651 commented on cs/TagsCloudVisualizationUnitTest/CircularCloudLayouterTest.cs:
```diff
@@ -0,0 +1,131 @@
+        [TestCase(10, TestName = "10 Rectangles")]
+        [TestCase(1000, TestName = "1000 Rectangles")]
+        [TestCase(2000, TestName = "2000 Rectangles")]
+        public void PutNextRectangle_ShouldBeOptimized(int countRectangle)
```
вижу что плохо усвоен предыдущий материал
я сейчас на какой нибудь малинке под доп нагрузкой запущу такой тест, что будет?

razor2651 commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,36 @@
+            nextAngle = 0;
+        }
+
+        public Point CalculatePointSpiral()
```
Spiral.CalculatePointSpiral()
опять дублирование(

razor2651 commented on cs/DebugProject/Program.cs:
```diff
@@ -0,0 +1,30 @@
+
+            painterOfRectangles.CreateImage(circularCloudLayouter.Rectangles, "CircularCloudLayouter1.png");
+
+            var p = new Process();
```
`p`
я такое еще раз увижу - сразу страйк кину

razor2651 reviewed (commented):
ты правишь код согласно замечаниям, но на следующей итерации делаешь ровно те же ошибки, пока что складывается ощущение что код пишется по принципу "на отвали"

я не знаю в чем конкретно проблема, возможно ты не придумал другого способа и решил сделать так... но если это так, то ты можешь всегда обратиться к наставнику, это человек который не только ревьюит код, но и может помочь если есть проблема

makspanshin committed:
Добавил интерфейсы

makspanshin committed:
Create SaverImageTest.cs

makspanshin committed:
Create PainterOfRectanglesTest.cs

makspanshin committed:
Убрал ненужный метод

makspanshin committed:
Удалил некорректный тест

Тесты на время делаются, если метод общается с внешнем миром(БД, сервисы и тд);

makspanshin committed:
Добавил вспомогательные методы

makspanshin committed:
Update CircularCloudLayouterTest.cs

makspanshin committed:
Create ArchimedesSpiralTest.cs

makspanshin committed:
Переименовал класс

makspanshin committed:
Create SaverImage.cs

makspanshin committed:
Update GeneratorOfRectangles.cs

makspanshin committed:
Добавил проверку на корректные размеры

makspanshin committed:
Изменил метод создания изображения

Выбрал такую реализацию, потому что класс работает с объектами наследующие IDisposable . Такими объекты опасно передавать и управлять ими должен тока хозяин.

makspanshin committed:
Update PainterOfRectangles.cs

makspanshin committed:
Update CircularCloudLayouter.cs

makspanshin committed:
Update Program.cs

razor2651 commented on cs/TagsCloudVisualization/PainterOfRectangles.cs:
```diff
@@ -0,0 +1,55 @@
+
+            using Image imageRectangles = new Bitmap(bmp);
+
+            command.Execute(imageRectangles);
```
ну ты по факту сделал даже еще хуже
я тебе в прошлый раз говорил что сохранению не место в логике рисовалки
ты же сейчас мало того что прорблему не исправил, дак еще и скрыл ее за абсолютно ничего ен говорящим интерфейсом
такого я еще не видел

razor2651 commented on cs/TagsCloudVisualization/Interfaces/IPainter.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace TagsCloudVisualization.Interfaces
+{
+    public interface IPainter
+    {
```
и что мне с таким интерфейсом делать?

окей, даже если ты сюда добавишь существующий

```
void CreateImage(List<Rectangle> rectangles, ICommandImage command)
```

я хочу написать вебапи которое бы отдавало картинку в ответ на запрос, мне как с таким методом работать? это нерасширяемый код

razor2651 commented on cs/TagsCloudVisualization/Interfaces/ICommandImage.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace TagsCloudVisualization.Interfaces
+{
+    public interface ICommandImage
```
что такое ICommandImage? он реализует всего один метод

```
void Execute(Image image);
```

я захочу сделать свою реализацию, видимо по твоей задумке я могу что угодно теперь делать с изображением... не ну а чо давай все методы называть теперь просто
`DoAnything(params object[] parameters)`

razor2651 commented on cs/TagsCloudVisualization/Interfaces/ISpiral.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public interface ISpiral
+    {
+        public Point CalculatePoint();
```
здесь указывать public не надо
если твой интерфейс и так public то и все методы в нем автоматически public если не указать чего то иного

razor2651 commented on cs/TagsCloudVisualization/SaverImage.cs:
```diff
@@ -0,0 +1,33 @@
+    {
+        private string filename;
+
+        public SaverImage(string filename)
```
то бишь если я хочу сохранить 2 файла то мне теперь надо еще и 2 инстанса этого saver делать?

razor2651 commented on cs/TagsCloudVisualizationUnitTest/PainterOfRectanglesTest.cs:
```diff
@@ -0,0 +1,52 @@
+        [Test]
+        public void CreateImage_ShouldBeThrow_When_CommandIsNull()
+        {
+            var painter = new PainterOfRectangles(new Size(10, 10));
```
ты в предыдущем классе почему то умеешь пользоваться setup чтобы избавиться от дублирования, а здесь уже разучился

razor2651 commented on cs/TagsCloudVisualizationUnitTest/PainterOfRectanglesTest.cs:
```diff
@@ -0,0 +1,52 @@
+        }
+
+        [Test]
+        public void CreateImage_ShouldBeThrow_When_CommandIsNull()
```
ты вообще в курсе что ты написал 2 теста одинаковых?

razor2651 reviewed (commented):
я не могу больше смотреть, это очень плохо, если на предыдущей итерации ревью мне только казалось что код пишется с пофигистическим отношением то ошибки которые я вижу сейчас только подтверждают эту теорию

с таким отношением надо что-то делать в дальнейшем
